pkg/stanza/fileconsumer/header.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline"
+)
+
+type HeaderConfig struct {
+	Pattern           string            `mapstructure:"pattern"`
+	MetadataOperators []operator.Config `mapstructure:"metadata_operators"`
+
+	// these are set by the "build" function
+	matchRegex *regexp.Regexp
+	splitFunc  bufio.SplitFunc
+}
+
+// validate returns an error describing why the configuration is invalid, or nil if the configuration is valid.
+func (hc *HeaderConfig) validate() error {
+	_, err := regexp.Compile(hc.Pattern)
+	if err != nil {
+		return fmt.Errorf("invalid `pattern`: %w", err)
+	}
+
+	if len(hc.MetadataOperators) == 0 {
+		return errors.New("at least one operator must be specified for `metadata_operators`")
+	}
+
+	nopLogger := zap.NewNop().Sugar()
+	outOp := newHeaderPipelineOutput(nopLogger)
+	p, err := pipeline.Config{
+		Operators:     hc.MetadataOperators,
+		DefaultOutput: outOp,
+	}.Build(nopLogger)
+
+	if err != nil {
+		return fmt.Errorf("failed to build pipelines: %w", err)
+	}
+
+	if !p.Operators()[0].CanProcess() {
+		return errors.New("first operator must be able to process entries")
+	}
+
+	return nil
+}
+
+func (hc *HeaderConfig) build(enc encoding.Encoding) error {
+	var err error
+	hc.matchRegex, err = regexp.Compile(hc.Pattern)
+	if err != nil {
+		return fmt.Errorf("failed to compile multiline pattern: %w", err)
+	}
+
+	hc.splitFunc, err = helper.NewNewlineSplitFunc(enc, false, func(b []byte) []byte {
+		return bytes.Trim(b, "\r\n")
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create split func: %w", err)
+	}
+
+	return nil
+}
+
+// buildHeader builds a header struct from the header config.
+func (hc *HeaderConfig) buildHeader(logger *zap.SugaredLogger) (*header, error) {
+	outOp := newHeaderPipelineOutput(logger)
+	p, err := pipeline.Config{
+		Operators:     hc.MetadataOperators,
+		DefaultOutput: outOp,
+	}.Build(logger)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to build pipeline: %w", err)
+	}
+
+	if err := p.Start(storage.NewNopClient()); err != nil {
+		return nil, fmt.Errorf("failed to start header pipeline: %w", err)
+	}
+
+	return &header{
+		config:               hc,
+		logger:               logger,
+		outputOperator:       outOp,
+		headerPipeline:       p,
+		attributesFromHeader: map[string]any{},
+	}, nil
+}
+
+type header struct {
+	config *HeaderConfig
+	logger *zap.SugaredLogger
+
+	finalized            bool
+	attributesFromHeader map[string]any
+	offset               int64
+
+	headerPipeline pipeline.Pipeline
+	outputOperator *headerPipelineOutput
+}
+
+// ReadHeader attempts to read the header from the given file. If the header is completed, fileAttributes will
+// have its HeaderAttributes field set to the resultant header attributes.
+func (h *header) ReadHeader(ctx context.Context, f io.ReadSeeker, maxLineSize int, enc helper.Encoding, fileAttributes *FileAttributes) {
+	if h.finalized {
+		return
+	}
```

Comment:
Are we expecting this case? From what I can see we are checking this before calling

Reply:
Good point, this is always checked by the caller.